otter/test/integration/test_lib.py

```diff
@@ -0,0 +1,137 @@
+"""
+Tests for the lib functions for convergence black-box testing.
+"""
+
+from twisted.internet.defer import Deferred
+from twisted.trial.unittest import SynchronousTestCase
+
+from otter.test.integration import lib
+
+
+class IdentityV2Tests(SynchronousTestCase):
+    """Tests for the :class:`IdentityV2` class."""
+
+    def test_missing_params(self):
+        """Missing parameters to the IdentityV2 constructor should raise
+        a ValueError exception.
+        """
+        def x(u, p, e):
+            self.assertRaises(ValueError, lib.IdentityV2, None, u, p, e)
+        x("", "pw", "ep")
+        x("un", "", "ep")
+        x("un", "pw", "")
+        x(None, "pw", "ep")
+        x("un", None, "ep")
+        x("un", "pw", None)
+
+    def test_pool_kwarg(self):
+        """The pool keyword argument should be passed through to the
+        authenticate_user Otter function.
+        """
+        rcs = lib.TestResources()
+
+        class Stub(object):
+            def __init__(self):
+                self.pool = None
+
+            def authenticate_user(self, endpt, user, passwd, pool=None):
+                self.pool = pool
+                return Deferred().addCallback(lambda: {})
+
+        stub = Stub()
+        lib.IdentityV2(
+            auth=stub, username="username",
+            password="password", endpoint="endpoint"
+        ).authenticate_user(rcs)
+        self.assertFalse(stub.pool)
+
+        lib.IdentityV2(
+            auth=stub, username="username",
+            password="password", endpoint="endpoint", pool=42
+        ).authenticate_user(rcs)
+        self.assertEquals(stub.pool, 42)
+
+    def test_records_results(self):
+        """The IdentityV2 instance you create should cache its
+        results when it receives them.
+        """
+        class Stub(object):
+            def authenticate_user(self, *unused_args, **unused_kwargs):
+                return Deferred().addCallback(lambda _: "cached")
```

Comment:
Same as above.  Here are equivalent statements to this:

- `return Deferred(); ... <d>.callback("cached")`
- `return succeed("cached")`  (without calling `.callback(...)` later)

The last one is generally preferred unless it's important for the Deferred to not return right away.